### DISCUSSION TITLE
How to reset part of the store synchronously at mount before first render?

### DISCUSSION THREAD

**Discussion by OoDeLally** [pmndrs/zustand#2648] [Q&A] [answered]
Let's say `MyPage` uses a `counter` that needs to be reset to `0` upon mount.
```tsx
const MyPage = (): ReactNode => {
  const { counter, resetCounter } = useStore();

  useEffect(() => {
    resetCounter();
  }, [resetCounter]);

  return <div>{counter}</div>;
};
```
Doing so would make `MyPage` render first with stale `counter` (from a previous state), and then re-render with a reset `counter`.

How can I avoid that first rerender, which seems like a glitch to me?

## Attempt 1: UseLayoutEffect

```tsx

const MyPage = (): ReactNode => {
  const { counter, resetCounter } = useStore();

  useLayoutEffect(() => {
    resetCounter();
  }, [resetCounter]);

  return <div>{counter}</div>;
};
```
Probably works, but is not recommended by React doc itself.

## Attempts 2: Synchronous set and ref

```ts

const MyPage = (): ReactNode => {
  const wasResetRef = useRef(false);
  const { resetCounter } = useStore();

  if (!wasResetRef.current) {
    useStore.getState().resetCounter();
    wasResetRef.current = true;
  }

  return <Inner />;
};

const Inner = () => {
  const { counter } = useStore();
  return <div>{counter}</div>;
};
```
But its ugly

What is a good way to achieve that?

Thanks for any idea!

**Comment by dai-shi** [comment]
> How can I avoid that first rerender, which seems like a glitch to me?

Hypothetically, you should reset the value before rendering `MyPage`. What could trigger rendering `MyPage`?

**Reply by OoDeLally** [reply]
Merely accessing the page.

**Reply by dai-shi** [reply]
So, window.onload? But then, the value has already been initial. Now I wonder if your use case fits with React Context usage, creating the store higher in the component tree instead of creating is globally.

**Reply by OoDeLally** [reply]
`window.onload` may execute when my page is not mounted yet. In addition, it will not trigger when the page is loaded a second time.

Traditionally for my needs, I would definitely use a `React.Context`, which can be initialized synchronously. But I also wish to have the granularity of store's state changes, as well as other nice features provided by zustand.

**Reply by dai-shi** [reply, marked as answer]
Sorry for the confusion. I meant using Context with Zustand, propagating the store itself.
https://github.com/pmndrs/zustand/blob/33cd0c0dd15307a98d859b7993c4160fa6f98b0b/docs/guides/initialize-state-with-props.md#basic-component-usage

**Reply by OoDeLally** [reply]
That's very nice, it seems to be exactly what I'm looking for. Thank you!
